# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом.



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Сделаем выборки семей без детей и с детьми:

In [ ]:
data_without_children = data.loc[data['children'] == 0]
data_with_children = data.loc[data['children'] > 0]

Теперь посчитаем среднее значение по столбцу debt для обоих групп населения:

In [ ]:
print('Без детей', round(data_without_children['debt'].mean(), 4))
print('С детьми', round(data_with_children['debt'].mean(), 4))

Без детей 0.0754
С детьми 0.0924


Как видно, есть разница между людьми с детьми и людьми без детей. Рассчитаем, насколько эти величины соотносятся между собой:

In [ ]:
round((1 - round(data_without_children['debt'].mean(), 4) / round(data_with_children['debt'].mean(), 4)) * 100, 1)

18.4

Получается, что у людей с детьми задолжностей по кредитам в среднем на 18,4 % больше, чем у людей без детей. Теперь посмотрим, как распределяются задолжности в зависимости от количества детей:

In [ ]:
round(data_with_children.groupby('children')['debt'].mean(), 4)

children
1    0.0923
2    0.0945
3    0.0818
4    0.0976
5    0.0000
Name: debt, dtype: float64

Прослеживается зависимость: чем больше детей, тем больше задолжностей по кредитам. Однако, тут есть аномальные значения:
1. У людей с тремя детьми задолжностей меньше, чем даже у людей с одним ребенком. Предполагаемое объяснение: случайное выпадающее значение.
2. У людей с пятью детьми задолжностей нет вообще. Возможное объяснение: малая выборка. Проверим это наблюдение:

In [ ]:
data_with_children.groupby('children')['debt'].count()

children
1    4808
2    2052
3     330
4      41
5       9
Name: debt, dtype: int64

Как и следовало ожидать. На всю выборку в 21525 строк всего 9 человек имеют 5 детей.

In [ ]:
data_pivot_children = data.pivot_table(index='children', values='debt', aggfunc='count')
data_pivot_children['numb_debt'] = data.loc[data['debt'] != 0].groupby('children')['debt'].count()  # число должников.
data_pivot_children['mean_debt'] = round(data.groupby('children')['debt'].mean(), 4)  # средняя доля должников в каждой категории людей.
data_pivot_children

debt  numb_debt  mean_debt
children                             
0         14091     1063.0     0.0754
1          4808      444.0     0.0923
2          2052      194.0     0.0945
3           330       27.0     0.0818
4            41        4.0     0.0976
5             9        NaN     0.0000

**Выводы:**
1. Люди с детьми в среднем допускают просрочку по кредиту на 18,4 % чаще, чем люди без детей.
2. Чем больше детей у человека, тем больше вероятность возникновения задолжности.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Посчитаем среднее значение задолжностей по каждому семейному статусу:

In [ ]:
data_family_debt_mean = data.groupby('family_status')['debt'].mean()
round(data_family_debt_mean.sort_values(), 4)

family_status
вдовец / вдова           0.0662
в разводе                0.0706
женат / замужем          0.0756
гражданский брак         0.0931
Не женат / не замужем    0.0976
Name: debt, dtype: float64

In [ ]:
data_pivot_family_status = data.pivot_table(index='family_status', values='debt', aggfunc='count')
data_pivot_family_status['numb_debt'] = data.loc[data['debt'] != 0].groupby('family_status')['debt'].count()  # число должников.
data_pivot_family_status['mean_debt'] = round(data.groupby('family_status')['debt'].mean(), 4)  # средняя доля должников в каждой категории людей.
data_pivot_family_status.sort_values(by='mean_debt')

debt  numb_debt  mean_debt
family_status                                     
вдовец / вдова           951         63     0.0662
в разводе               1189         84     0.0706
женат / замужем        12261        927     0.0756
гражданский брак        4134        385     0.0931
Не женат / не замужем   2796        273     0.0976

Из расчета видно, что наименьшее число задолжностей у людей, которые вдовцы/вдовы. Возможно, это связано с возрастом. Проверим эту гипотезу:

In [ ]:
data_age_group_mean = data.groupby('family_status')['dob_years'].mean()
round(data_age_group_mean.sort_values(), 2)

family_status
Не женат / не замужем    38.37
гражданский брак         42.07
женат / замужем          43.56
в разводе                45.56
вдовец / вдова           56.50
Name: dob_years, dtype: float64

Гипотеза подтвердилась. Вдовцы и вдовы имеют высокий средний возраст.

**Вывод:** Вдовцы/вдовы наиболее надежные заемщики.
Наименее надежные заемщики - неженатые люди.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Посчитаем среднее значение задолжностей по каждой категории доходов:

In [ ]:
data_income_debt_mean = data.groupby('total_income_category')['debt'].mean()
round(data_income_debt_mean.sort_values(), 4)

total_income_category
D    0.0602
B    0.0706
A    0.0800
C    0.0850
E    0.0909
Name: debt, dtype: float64

In [ ]:
data_pivot_total_income = data.pivot_table(index='total_income_category', values='debt', aggfunc='count')
data_pivot_total_income['numb_debt'] = data.loc[data['debt'] != 0].groupby('total_income_category')['debt'].count()  # число должников.
data_pivot_total_income['mean_debt'] = round(data.groupby('total_income_category')['debt'].mean(), 4)  # средняя доля должников в каждой категории людей.
data_pivot_total_income

debt  numb_debt  mean_debt
total_income_category                             
A                         25          2     0.0800
B                       5014        354     0.0706
C                      15921       1353     0.0850
D                        349         21     0.0602
E                         22          2     0.0909

**Вывод:** в категории А и категории Е (очень высокий и очень низкий доход) людей очень мало. В связи с этим, выводы сделать затруднительно. Можно лишь сообщить, что у людей со средним доходом доля просрочек по кредитам выше, чем у людей с доходом выше среднего.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Посчитаем среднее значение задолжностей по целям, на которые берется кредит:

In [ ]:
data_purpose_debt_mean = data.groupby('purpose_category')['debt'].mean()
round(data_purpose_debt_mean.sort_values(), 4)

purpose_category
операции с недвижимостью    0.0726
проведение свадьбы          0.0791
получение образования       0.0925
операции с автомобилем      0.0935
Name: debt, dtype: float64

In [ ]:
data_pivot_purpose = data.pivot_table(index='purpose_category', values='debt', aggfunc='count')
data_pivot_purpose['numb_debt'] = data.loc[data['debt'] != 0].groupby('purpose_category')['debt'].count()  # число должников.
data_pivot_purpose['mean_debt'] = round(data.groupby('purpose_category')['debt'].mean(), 4)  # средняя доля должников в каждой категории людей.
data_pivot_purpose.sort_values(by='mean_debt')

debt  numb_debt  mean_debt
purpose_category                                     
операции с недвижимостью  10751        780     0.0726
проведение свадьбы         2313        183     0.0791
получение образования      3988        369     0.0925
операции с автомобилем     4279        400     0.0935

**Вывод:** Наименьшее число задолжностей у людей, которые взяли кредит на недвижимость, наибольшее - на цели, связанные с автомобилем.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски были в столбцах в days_employed (количество дней занятости) и в total_income (общий доход). Наиболее вероятная причина пропусков: у человека нет официальной работы (занятости) в связи с этим нет и официальных доходов.  
Стоит обратить внимание, что в таблице возможно отсутствует информация о пенсиях и других социальных выплатах. Так, на строке 12 изначального DataFrame (Задание 2) есть человек со статусом "Пенсионер", у которого нет ни времени занятости, ни доходов, но возраст указан 65 лет. Маловероятно, что за всю жизнь у него не было работы. В этом случае кажется вероятным, что значения не поступили из нужной базы либо сам человек не счел нужным их сообщить.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В данном случае медианное значение наиболее подходящее, так как могут быть как очень высокие (или очень низкие) значения доходов, так и очень высокие (или очень низние) значения занятости. В связи с этим среднее значение не будет объективно, так как оно будет смещено либо вверх либо вниз.

### Шаг 4: общий вывод.

1. Представлены исходные данные в количестве 21525 строк. В 2174 строке были обнаружены пропущенные значения. Пропущенные значения были в стоблцах "days_employed" (количество дней занятости) и в "total_income" (общий доход). Пустые значения были заменены на медианные в целях повышения точности анализа.  
Также были обнаружены аномальные значения в столбце "children" (дети): 20 и -1. Указанные значения были удалены как аномальные.  
Для удобства анализа общие доходы и цели кредитов были разделены на категории.


2. В общем случае на вероятность возникновения просрочки по кредиту влияет:
- семейное положение человека. Так, среди вдовцов/вдов наблюдается наименьшая доля просрочек.
- цель кредита. Наименьшая доля просрочек у кредитов, выданных для операций с недвижимостью.
- наличие или отсутствует детей. У людей с детьми доля просрочек больше, чем у людей без детей.  
- На основе представленных данных нет возможности сделать вывод о связи между доходом человека и вероятностью возникновения задолжности.


3. В целях уточнения анализа предлагается:
- дополнить данные по доходам из других источников, к примеру, по данным различных фондов социальной поддержки.
- собрать данные по величине задержек по кредитам у людей с высоким доходом и с низким доходом. Возможно, указанные данные могут сделать выводы о зависимости доходов и вероятности просрочки по кредиту.